test/node.go

```diff
@@ -0,0 +1,567 @@
+func NewNode(c *NodeConfig, genesis *core.Genesis) (*Node, error) {
+
+	// p2p key and address
+	c.P2P.PrivateKey = c.ValidatorAccount.PrivateKey
+
+	// Make temp datadir
+	datadir, err := ioutil.TempDir("", "celo_datadir")
+	if err != nil {
+		return nil, err
+	}
+	c.DataDir = datadir
+
+	// copy the base eth config, so we can modify it without damaging the
+	// original.
+	ec := &eth.Config{}
+	err = copyObject(baseEthConfig, ec)
+	if err != nil {
+		return nil, err
+	}
+	ec.Genesis = genesis
+	ec.NetworkId = genesis.Config.ChainID.Uint64()
+
+	node := &Node{
+		Config:     c.Config,
+		EthConfig:  ec,
+		Key:        c.ValidatorAccount.PrivateKey,
+		Address:    c.ValidatorAccount.Address,
+		DevAddress: c.DevAccount.Address,
+		DevKey:     c.DevAccount.PrivateKey,
+		Tracker:    NewTransactionTracker(),
+	}
+
+	return node, node.Start()
+}
+
+// Start creates the node.Node and eth.Ethereum and starts the node.Node and
+// starts eth.Ethereum mining.
+func (n *Node) Start() error {
+	// Provide a copy of the config to node.New, so that we can rely on
+	// Node.Config field not being manipulated by node and hence use our copy
+	// for black box testing.
+	nodeConfigCopy := &node.Config{}
+	err := copyNodeConfig(n.Config, nodeConfigCopy)
+	if err != nil {
+		return err
+	}
+
+	// Give this logger context based on the node address so that we can easily
+	// trace single node execution in the logs. We set the logger only on the
+	// copy, since it is not useful for black box testing and it is also not
+	// marshalable since the implementation contains unexported fields.
+	//
+	// Note unfortunately there are many other loggers created in geth separate
+	// from this one, which means we still see a lot of output that is not
+	// attributable to a specific node.
+	nodeConfigCopy.Logger = log.New("node", n.Address.String()[2:7])
+	nodeConfigCopy.Logger.SetHandler(log.LvlFilterHandler(log.LvlTrace, log.StreamHandler(os.Stdout, log.TerminalFormat(true))))
+
+	n.Node, err = node.New(nodeConfigCopy)
+	if err != nil {
+		return err
+	}
+
+	// This registers the ethereum service on n.Node, so that calling
+	// n.Node.Stop will also close the eth service. Again we provide a copy of
+	// the EthConfig so that we can use our copy for black box testing.
+	ethConfigCopy := &eth.Config{}
+	err = copyObject(n.EthConfig, ethConfigCopy)
+	if err != nil {
+		return err
+	}
+	ethConfigCopy.Miner.Validator = n.Address
+	ethConfigCopy.TxFeeRecipient = n.Address
+	ethConfigCopy.Istanbul.AnnounceQueryEnodeGossipPeriod = 60
+	ethConfigCopy.Istanbul.Validator = true
+	// ethConfigCopy.Istanbul.BlockPeriod = 0
+	// Somehow this is not being honored, its being set to 3000
+	// ethConfigCopy.Istanbul.RequestTimeout = 100
+	ethConfigCopy.Istanbul.TimeoutBackoffFactor = 10
+	ethConfigCopy.Istanbul.MinResendRoundChangeTimeout = 15 * 1000
+	ethConfigCopy.Istanbul.MaxResendRoundChangeTimeout = 2 * 60 * 1000
+	// ethConfigCopy.Istanbul.Epoch = 100
```

Comment:
dead code or a future todo?
ditto for the above commented out code.

Reply:
Just some dev cruft, I think they can be safely removed.